File: packages/senri_ffi_js/src/types/normalized.ts
/*
 * Copyright (c) 2026 TT23XR Studio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { TYPES, TYPE_SIZES, TYPE_ALIGNS } from './constants';
import { FFITypeError } from '../errors';

export type PrimitiveName =
  | 'void'
  | 'int8' | 'uint8'
  | 'int16' | 'uint16'
  | 'int32' | 'uint32'
  | 'int64' | 'uint64'
  | 'float32' | 'float64'
  | 'cstring'
  | 'pointer';

export type NormalizedType =
  | { kind: 'primitive'; name: PrimitiveName }
  | { kind: 'pointer'; of: NormalizedType }
  | { kind: 'array'; of: NormalizedType; length: number }
  | { kind: 'struct'; fields: Record<string, NormalizedType>; packed?: number; size: number; align: number };

export function getNormalizedSize(type: NormalizedType): number {
  switch (type.kind) {
    case 'primitive': return TYPE_SIZES[type.name];
    case 'pointer':   return TYPE_SIZES.pointer;
    case 'array':     return getNormalizedSize(type.of) * type.length;
    case 'struct':    return type.size;
  }
}

export function getNormalizedAlign(type: NormalizedType): number {
  switch (type.kind) {
    case 'primitive': return TYPE_ALIGNS[type.name];
    case 'pointer':   return TYPE_ALIGNS.pointer;
    case 'array':     return getNormalizedAlign(type.of);
    case 'struct':    return type.align;
  }
}

function isTypeDescriptor(t: any): boolean {
  return t && typeof t === 'object' && typeof t.__senri_type === 'string';
}

function normalizeFields(fields: Record<string, any>): Record<string, NormalizedType> {
  const result: Record<string, NormalizedType> = {};
  for (const [name, ft] of Object.entries(fields)) {
    result[name] = normalizeType(ft);
  }
  return result;
}

function isPrimitiveString(t: any): t is string {
  return typeof t === 'string' && t in TYPES;
}

export function normalizeType(type: any): NormalizedType {
  if (isPrimitiveString(type)) {
    return { kind: 'primitive', name: type as PrimitiveName };
  }

  if (!isTypeDescriptor(type)) {
    throw new FFITypeError('Unknown type descriptor: ' + JSON.stringify(type));
  }

  const tt = type.__senri_type;

  if (tt === 'pointer') {
    return {
      kind: 'pointer',
      of: normalizeType(type.innerType || TYPES.pointer),
    };
  }

  if (tt === 'array') {
    return {
      kind: 'array',
      of: normalizeType(type.innerType),
      length: type.length || 0,
    };
  }

  if (tt === 'struct') {
    const fields = normalizeFields(type.fields);
    const fieldEntries: Record<string, any> = {};
    for (const [name, nt] of Object.entries(fields)) {
      fieldEntries[name] = nt;
    }
    const totalSize = computeNormalizedLayout(fieldEntries, type.packed);
    let maxAlign = 1;
    for (const nt of Object.values(fields)) {
      maxAlign = Math.max(maxAlign, getNormalizedAlign(nt));
    }
    return {
      kind: 'struct',
      fields,
      packed: type.packed,
      size: totalSize,
      align: maxAlign,
    };
  }

  throw new FFITypeError('Unknown type descriptor: ' + JSON.stringify(type));
}

function computeNormalizedLayout(
  fields: Record<string, NormalizedType>,
  packed?: number
): number {
  let offset = 0;
  let maxAlign = 1;

  for (const nt of Object.values(fields)) {
    const align = packed ? Math.min(getNormalizedAlign(nt), packed) : getNormalizedAlign(nt);
    if (!packed && offset % align !== 0) {
      offset = Math.ceil(offset / align) * align;
    }
    offset += getNormalizedSize(nt);
    maxAlign = Math.max(maxAlign, align);
  }

  return packed ? offset : Math.ceil(offset / maxAlign) * maxAlign;
}

export function serializeType(type: NormalizedType): string {
  switch (type.kind) {
    case 'primitive': return 'p:' + type.name;
    case 'pointer':   return '*' + serializeType(type.of);
    case 'array':     return '[' + type.length + ']' + serializeType(type.of);
    case 'struct': {
      const fields = Object.entries(type.fields)
        .map(([k, v]) => k + '=' + serializeType(v))
        .sort()
        .join(',');
      return '{' + (type.packed ? '#' : '') + fields + '}';
    }
  }
}

File: packages/senri_ffi_py/src/senri_ffi/types/normalized.py
from __future__ import annotations

from typing import Any, Literal

from senri_ffi.errors import FFITypeError
from senri_ffi.types.constants import TYPES, TYPE_SIZES, TYPE_ALIGNS

PrimitiveName = Literal[
    "void",
    "int8", "uint8",
    "int16", "uint16",
    "int32", "uint32",
    "int64", "uint64",
    "float32", "float64",
    "cstring",
    "pointer",
]


class NormalizedType:
    def __init__(
        self,
        kind: Literal["primitive", "pointer", "array", "struct"],
        name: PrimitiveName | None = None,
        of: NormalizedType | None = None,
        length: int = 0,
        fields: dict[str, NormalizedType] | None = None,
        packed: int | None = None,
        size: int = 0,
        align: int = 1,
    ) -> None:
        self.kind = kind
        self.name = name
        self.of = of
        self.length = length
        self.fields = fields
        self.packed = packed
        self.size = size
        self.align = align

    @classmethod
    def primitive(cls, name: PrimitiveName) -> NormalizedType:
        return cls(kind="primitive", name=name)

    @classmethod
    def pointer(cls, of: NormalizedType) -> NormalizedType:
        return cls(kind="pointer", of=of)

    @classmethod
    def array(cls, of: NormalizedType, length: int) -> NormalizedType:
        return cls(kind="array", of=of, length=length)

    @classmethod
    def struct(
        cls,
        fields: dict[str, NormalizedType],
        packed: int | None = None,
        size: int = 0,
        align: int = 1,
    ) -> NormalizedType:
        return cls(kind="struct", fields=fields, packed=packed, size=size, align=align)

    def __repr__(self) -> str:
        if self.kind == "primitive":
            return f"NormalizedType({self.kind}, {self.name})"
        if self.kind == "pointer":
            return f"NormalizedType({self.kind}, of={self.of})"
        if self.kind == "array":
            return f"NormalizedType({self.kind}, of={self.of}, length={self.length})"
        return f"NormalizedType({self.kind}, fields={{{list(self.fields or {})}}})"


def get_normalized_size(t: NormalizedType) -> int:
    if t.kind == "primitive":
        assert t.name is not None
        return TYPE_SIZES[t.name]
    if t.kind == "pointer":
        return TYPE_SIZES["pointer"]
    if t.kind == "array":
        assert t.of is not None
        return get_normalized_size(t.of) * t.length
    return t.size


def get_normalized_align(t: NormalizedType) -> int:
    if t.kind == "primitive":
        assert t.name is not None
        return TYPE_ALIGNS[t.name]
    if t.kind == "pointer":
        return TYPE_ALIGNS["pointer"]
    if t.kind == "array":
        assert t.of is not None
        return get_normalized_align(t.of)
    return t.align


def _is_type_descriptor(t: Any) -> bool:
    if isinstance(t, dict) and "__senri_type" in t:
        return True
    if isinstance(t, type) and hasattr(t, "__senri_type"):
        return True
    return False


def _is_primitive_string(t: Any) -> bool:
    return isinstance(t, str) and t in TYPES


def normalize_fields(fields: dict[str, Any]) -> dict[str, NormalizedType]:
    return {k: normalize_type(v) for k, v in fields.items()}


def compute_normalized_layout(
    fields: dict[str, NormalizedType],
    packed: int | None = None,
) -> int:
    offset = 0
    max_align = 1
    for nt in fields.values():
        if packed is not None:
            a = min(get_normalized_align(nt), packed)
        else:
            a = get_normalized_align(nt)
        if not packed and offset % a != 0:
            offset = ((offset // a) + 1) * a
        offset += get_normalized_size(nt)
        max_align = max(max_align, a)
    if packed:
        return offset
    return ((offset + max_align - 1) // max_align) * max_align


def normalize_type(t: Any) -> NormalizedType:
    if _is_primitive_string(t):
        return NormalizedType.primitive(t)

    if not _is_type_descriptor(t):
        msg = f"Unknown type descriptor: {t!r}"
        raise FFITypeError(msg)

    if isinstance(t, type):
        tt = getattr(t, "__senri_type", None)
    else:
        tt = t.get("__senri_type") if isinstance(t, dict) else getattr(t, "__senri_type", None)

    if tt == "pointer":
        if isinstance(t, type):
            raise FFITypeError(f"Cannot normalize pointer from type object: {t!r}")
        inner = normalize_type(t.get("innerType", TYPES["pointer"]))
        return NormalizedType.pointer(inner)

    if tt == "array":
        if isinstance(t, type):
            raise FFITypeError(f"Cannot normalize array from type object: {t!r}")
        inner = normalize_type(t["innerType"])
        return NormalizedType.array(inner, t.get("length", 0))

    if tt == "struct":
        fields_dict = t.fields if isinstance(t, type) else t.get("fields", {})
        fields = normalize_fields(fields_dict)
        packed = t.packed if isinstance(t, type) else t.get("packed")
        total_size = compute_normalized_layout(fields, packed)
        max_align = 1
        for nt in fields.values():
            max_align = max(max_align, get_normalized_align(nt))
        return NormalizedType.struct(fields, packed, total_size, max_align)

    msg = f"Unknown type descriptor: {t!r}"
    raise FFITypeError(msg)


def serialize_type(t: NormalizedType) -> str:
    if t.kind == "primitive":
        assert t.name is not None
        return "p:" + t.name
    if t.kind == "pointer":
        assert t.of is not None
        return "*" + serialize_type(t.of)
    if t.kind == "array":
        assert t.of is not None
        return f"[{t.length}]" + serialize_type(t.of)
    if t.kind == "struct":
        parts = sorted(f"{k}={serialize_type(v)}" for k, v in (t.fields or {}).items())
        prefix = "#" if t.packed else ""
        return "{" + prefix + ",".join(parts) + "}"
    return ""
